### PR TITLE
cabal: Add version constraint on random

### LINKED ISSUE #2685: Cabal build fails with random-1.2.0
# Context

When trying to compile `cardano-wallet` using cabal I'm getting the following error:

```
src/Cardano/Wallet/DB/Sqlite/Types.hs:575:24: error:
    • No instance for (Read StdGen)
        arising from a use of ‘fromPersistValueRead’
    • In the expression: fromPersistValueRead
      In an equation for ‘fromPersistValue’:
          fromPersistValue = fromPersistValueRead
      In the instance declaration for ‘PersistField StdGen’
    |
575 |     fromPersistValue = fromPersistValueRead
    |                        ^^^^^^^^^^^^^^^^^^^^
cabal: Failed to build cardano-wallet-core-2021.4.28 (which is required by
exe:local-cluster from cardano-wallet-2021.4.28, test:integration from
cardano-wallet-2021.4.28 and others).
```

tried ghc `8.10.2` and `8.10.4`

## Essential Information

 - **Version**:      `cardano-wallet` release/tag ` v2021-05-26`
 - **Platform**:    Linux (Ubuntu 20.04)
 - **Installation**: from source
 - **Network**:  any

# Steps to Reproduce
1. git clone https://github.com/input-output-hk/cardano-wallet.git
2. git fetch --all --recurse-submodules --tags
3. git checkout tags/v2021-05-26
4. cabal configure --with-compiler=ghc-8.10.4
5. cabal build all

## Expected behavior

I expect it to compile succesfully

## Actual behavior

I get the following error and compilation fails:

```
src/Cardano/Wallet/DB/Sqlite/Types.hs:575:24: error:
    • No instance for (Read StdGen)
        arising from a use of ‘fromPersistValueRead’
    • In the expression: fromPersistValueRead
      In an equation for ‘fromPersistValue’:
          fromPersistValue = fromPersistValueRead
      In the instance declaration for ‘PersistField StdGen’
    |
575 |     fromPersistValue = fromPersistValueRead
    |                        ^^^^^^^^^^^^^^^^^^^^
cabal: Failed to build cardano-wallet-core-2021.4.28 (which is required by
exe:local-cluster from cardano-wallet-2021.4.28, test:integration from
cardano-wallet-2021.4.28 and others).
```

# Jira ticket

ADP-961

### PR DESCRIPTION
**Pull request by rvl**
# Issue Number

ADP-961

# Overview

In `random>=1.2.0`, `StdGen` no longer has a `Read` instance.

- Add version constraint so Cabal uses previous version.
- Clean up `cabal.project`, with the assumption of Cabal >= 3.4.0.0.
- Add a nightly CI step to run the cabal build.
- Use cabal-cache to make the nightly build faster.
- Make sure every Haskell tool built for the nix-shell has a materialization, for faster evaluation without warnings.

# Comments

Resolves #2685.

We may fix the usage of `instance Read StdGen` later.

### CODE REVIEW

**Review by sevanspowell — APPROVED** [review]

File: cabal.project [hunk 1]
```diff
@@ -178,37 +178,9 @@ package cardano-wallet-test-utils
 package strict-non-empty-containers
   tests: True
 
--- Disable all other tests. This is what they do in the cardano-node
--- cabal project.
--- The global tests: False flag doesn't seem to work though.
+-- Now disable all other tests with a global flag.
```
Comment (sevanspowell) [inline cabal.project:181]: Does this actually work? I've always had trouble with cabal.project's and global flags. 

I've tried things like

```
package *:
  tests: False
```

In the past but haven't had it work.

Comment (sevanspowell) [reply cabal.project:181]: Always confuses me.

Comment (rvl) [reply cabal.project:181]: It seems to work for me locally, and the CI builds also work (whereas before they were failing due to difficult test suite dependencies). I think key part is using Cabal 3.4.0.0.

**Pull request merged**